packages/editor/src/store/actions.js

```diff
@@ -910,28 +910,27 @@ export function* resetEditorBlocks( blocks, options = {} ) {
 	// Sync to sources from block attributes updates.
 	if ( lastBlockAttributesChange ) {
 		const updatedSources = new Set;
-
-		const blockStack = [ ...blocks ];
-
-		for ( let i = 0; i < blockStack.length; i++ ) {
-			const block = blockStack[ i ];
-
-			// Account for changes occurring within nested block.
-			if ( block.innerBlocks.length ) {
-				blockStack.push( ...block.innerBlocks );
-			}
-
-			if ( ! lastBlockAttributesChange.hasOwnProperty( block.clientId ) ) {
+		const updatedBlockTypes = new Set;
+		for ( const [ clientId, attributes ] of Object.entries( lastBlockAttributesChange ) ) {
+			const blockName = yield select( 'core/block-editor', 'getBlockName', clientId );
+			if ( updatedBlockTypes.has( blockName ) ) {
 				continue;
 			}
 
-			const blockType = yield select( 'core/blocks', 'getBlockType', block.name );
-			const changedAttributes = lastBlockAttributesChange[ block.clientId ];
+			updatedBlockTypes.add( blockName );
+			const blockType = yield select( 'core/blocks', 'getBlockType', blockName );
 
 			for ( const [ attributeName, schema ] of Object.entries( blockType.attributes ) ) {
```

Comment:
Nice job on this optimization!

It would also be more intuitive to just loop over `attributes` here.

Reply:
> It would also be more intuitive to just loop over `attributes` here.

Agreed! This is another optimization, since the updated attributes are likely a subset of the attributes of a block type (and most likely just a single attribute), so we can avoid iterating over all attributes of a block type by arranging it this way.

I've revised this in 38d27e1.